src/replication.c

```diff
@@ -2160,136 +2182,131 @@ void syncWithMaster(connection *conn) {
             strncmp(err,"-ERR operation not permitted",28) != 0)
         {
             serverLog(LL_WARNING,"Error reply to PING from master: '%s'",err);
-            sdsfree(err);
+            sdsfree(err), err = NULL;
             goto error;
         } else {
             serverLog(LL_NOTICE,
                 "Master replied to PING, replication can continue...");
         }
-        sdsfree(err);
-        server.repl_state = REPL_STATE_SEND_AUTH;
+        sdsfree(err), err = NULL;
+        server.repl_state = REPL_STATE_SEND_HANDSHAKE;
     }
 
-    /* AUTH with the master if required. */
-    if (server.repl_state == REPL_STATE_SEND_AUTH) {
+    if (server.repl_state == REPL_STATE_SEND_HANDSHAKE) {
+        /* AUTH with the master if required. */
         if (server.masterauth) {
-            sds auth = sdsnew("AUTH");
+            char *args[3] = {"AUTH",NULL,NULL};
+            size_t lens[3] = {4,0,0};
+            int argc = 1;
             if (server.masteruser) {
-                sds masteruser = sdsnew(server.masteruser);
-                err = sendSynchronousCommand(SYNC_CMD_WRITE | SYNC_CMD_WRITE_SDS, conn, auth,
-                                             masteruser, server.masterauth, NULL);
-                sdsfree(masteruser);
-            } else {
-                err = sendSynchronousCommand(SYNC_CMD_WRITE | SYNC_CMD_WRITE_SDS, conn, auth,
-                                             server.masterauth, NULL);
+                args[argc] = server.masteruser;
+                lens[argc] = strlen(server.masteruser);
+                argc++;
             }
-            sdsfree(auth);
+            args[argc] = server.masterauth;
+            lens[argc] = sdslen(server.masterauth);
+            argc++;
+            err = sendCommandArgv(conn, argc, args, lens);
             if (err) goto write_error;
-            server.repl_state = REPL_STATE_RECEIVE_AUTH;
-            return;
-        } else {
-            server.repl_state = REPL_STATE_SEND_PORT;
         }
+
+        /* Set the slave port, so that Master's INFO command can list the
```

Comment:
Consider breaking this (and possible more) into smaller functions instead of just extra scoping?

Reply:
this function will not be used anywhere else. and i'm not sure extracting this into a function is justified due to the "clear role and input/output" concern.
in this case although this is a huge function, i don't think that taking each of these 5 line blocks into a separate function makes sense. (they're only used in this context)
are you just concerned about the variable declaration? redis actually has many places today that declare variables mid function without opening a scope. (and also we can move them to the top of the function)

Reply:
Not concerned about the variables declaration, only about the length of the function and breaking it down into named segments. Another benefit is it's one step closer to refactoring everything to be really async.